db.go

```diff
@@ -246,6 +251,145 @@ func newDBMetrics(db *DB, r prometheus.Registerer) *dbMetrics {
+func OpenDBReadOnly(dir string, l log.Logger) (*DBReadOnly, error) {
+	if _, err := os.Stat(dir); err != nil {
+		return nil, err
+	}
+
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	db := &DBReadOnly{
+		logger: l,
+		dir:    dir,
+	}
+
+	return db, nil
+}
+
+// Querier loads the wal and returns a new querier over the data partition for the given time range.
+// A goroutine must not handle more than one open Querier.
+func (db *DBReadOnly) Querier(mint, maxt int64) (Querier, error) {
+	w, err := wal.NewReadOnly(db.logger, nil, filepath.Join(db.dir, "wal"))
+	if err != nil {
+		return nil, err
+	}
+	head, err := NewHead(nil, db.logger, w, 1)
+	if err != nil {
+		return nil, err
+	}
+
+	blocksReadOnly, err := db.Blocks()
+	if err != nil {
+		return nil, err
+	}
+
+	blocks := make([]*Block, len(blocksReadOnly))
+	for i, b := range blocksReadOnly {
+		b, ok := b.(*Block)
+		if !ok {
+			return nil, errors.New("unable to convert a read only block to a normal block")
+		}
+		blocks[i] = b
+	}
+
+	dbWritable := &DB{
+		dir:    db.dir,
+		logger: db.logger,
+		head:   head,
+		blocks: blocks,
+	}
+
+	// Set the min valid time for the ingested wal samples
+	// to be no lower than the maxt of the last block.
+	minValidTime := int64(math.MinInt64)
+	if len(blocks) > 0 {
+		minValidTime = blocks[len(blocks)-1].Meta().MaxTime
+	}
+
+	if err := dbWritable.head.Init(minValidTime); err != nil {
+		return nil, errors.Wrap(err, "read WAL")
+	}
+
+	return dbWritable.Querier(mint, maxt)
+}
+
+// Blocks returns all persisted blocks.
+func (db *DBReadOnly) Blocks() ([]BlockReadOnly, error) {
+	loadable, corrupted, err := openBlocks(db.logger, db.dir, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Corrupted blocks that have been superseded by a loadable block can be safely ignored.
+	for _, block := range loadable {
+		for _, b := range block.Meta().Compaction.Parents {
+			delete(corrupted, b.ULID)
+		}
+	}
+	if len(corrupted) > 0 {
+		for _, b := range loadable {
+			if err := b.Close(); err != nil {
+				level.Warn(db.logger).Log("msg", "closing a block", err)
+			}
+		}
+		return nil, fmt.Errorf("unexpected corrupted block:%v", corrupted)
+	}
+
+	if len(loadable) == 0 {
+		return nil, errors.New("no blocks found")
+	}
+
+	sort.Slice(loadable, func(i, j int) bool {
+		return loadable[i].Meta().MinTime < loadable[j].Meta().MinTime
+	})
+
+	blockMetas := make([]BlockMeta, 0, len(loadable))
+	for _, b := range loadable {
+		blockMetas = append(blockMetas, b.Meta())
+	}
+	if overlaps := OverlappingBlocks(blockMetas); len(overlaps) > 0 {
+		level.Warn(db.logger).Log("msg", "overlapping blocks found during opening", "detail", overlaps.String())
+	}
+
+	// Close all previously open blocks and add the new ones to the catche.
+	for _, b := range db.blocks {
+		b.Close()
+	}
+
+	blocks := make([]BlockReadOnly, len(loadable))
+	for i, b := range loadable {
+		blocks[i] = b
+	}
+	db.blocks = blocks
```

Comment:
For the current use case, this looks fine. We can run into race conditions if it used concurrently by multiple goroutines in the future (more like wherever we are using `db.blocks`). We can address this in the future if needed, just adding a note here.

Reply:
added a lock